llvm/lib/Transforms/IPO/MergeFunctions.cpp

```diff
@@ -375,9 +375,35 @@ bool MergeFunctions::doFunctionalCheck(std::vector<WeakTrackingVH> &Worklist) {
 }
 #endif
 
+// This function iterates over two functions and compares their
+// metadata to make sure we aren't merging functions which have
+// distinct metadata that is different. We don't care about
```

Comment:
I think this should be more precise. It's not actually about "having" metadata. Many many functions will have metadata *attachments* on lots of instructions, which may be distinct without any trouble. The problem is only when an intrinsic references distinct metadata as an operand.